packages/flame_3d/lib/src/parser/gltf/gltf_root.dart

```diff
@@ -0,0 +1,156 @@
+  late final List<Camera> cameras;
+  late final List<Skin> skins;
+  late final List<Mesh> meshes;
+  late final List<Material> materials;
+  late final List<Texture> textures;
+  late final List<Animation> animations;
+  late final List<Sampler> samplers;
+  late final List<Image> images;
+
+  late final List<GlbChunk> chunks;
+
+  GltfRoot._({
+    required String prefix,
+  }) : _prefix = prefix;
+
+  Future<Uint8List> readChunk(GltfRef<Buffer> ref) async {
+    if (chunks.isNotEmpty) {
+      final chunk = chunks[ref.index];
+      return chunk.data;
+    }
+    final buffer = ref.get();
+    return readChunkFrom(buffer.uri!);
+  }
+
+  Future<Uint8List> readChunkFrom(String uri) async {
+    if (uri.startsWith('data:')) {
+      const prefixes = [
+        'data:application/gltf-buffer;base64,',
+        'data:application/octet-stream;base64,',
+      ];
+      for (final prefix in prefixes) {
+        if (uri.startsWith(prefix)) {
+          return base64Decode(uri.substring(prefix.length));
+        }
+      }
+      throw Exception('Unsupported data URI: $uri');
+    } else {
+      final path = '$_prefix/$uri';
+      return Flame.assets.readBinaryFile(path);
+    }
+  }
+
+  T resolve<T extends GltfNode>(int index) {
+    return switch (T) {
+      const (Scene) => scenes[index],
+      const (Node) => nodes[index],
+      const (Mesh) => meshes[index],
+      const (Material) => materials[index],
+      const (Camera) => cameras[index],
+      const (Skin) => skins[index],
+      const (BufferView) => bufferViews[index],
+      const (Buffer) => buffers[index],
+      const (Texture) => textures[index],
+      const (Animation) => animations[index],
+      const (Sampler) => samplers[index],
+      const (Image) => images[index],
+      const (IntAccessor) => accessors[index].asInt(),
+      const (FloatAccessor) => accessors[index].asFloat(),
+      const (Vector2Accessor) => accessors[index].asVector2(),
+      const (Vector3Accessor) => accessors[index].asVector3(),
+      const (Vector4Accessor) => accessors[index].asVector4(),
+      const (QuaternionAccessor) => accessors[index].asQuaternion(),
+      const (Matrix4Accessor) => accessors[index].asMatrix4(),
+      const (RawAccessor) => accessors[index],
+      _ => throw UnimplementedError('Cannot resolve type $T')
+    } as T;
+  }
+
+  static Future<GltfRoot> from({
+    required String prefix,
+    required Map<String, dynamic> json,
+    required List<GlbChunk> chunks,
+  }) async {
+    final root = GltfRoot._(prefix: prefix);
+    root.chunks = chunks;
+
+    Future<List<T>> parse<T>(
+      String key,
+      T Function(GltfRoot, Map<String, Object?>) parser,
+    ) async {
+      final objects = Parser.objectList(root, json, key, parser) ?? [];
+      for (final object in objects) {
+        if (object is GltfNodeWithData) {
+          await object.init();
+        }
+      }
+      return objects;
+    }
+
+    root.buffers = await parse('buffers', Buffer.parse);
```

Comment:
Maybe all this parsing should be done in parallel? Or can init contain things that can be parallelized? Otherwise it's no gain to parallelizing.

Reply:
i'd be afraid of prematurely optmize it and cause concurrency issues, not sure if all the maps and stuff used are safe to be accessed concurrently. since this is only done at game load, I'd rather leave that for a followup to be considered with more care